### PR TITLE
feat(discover): Implement download as CSV

### PR DESCRIPTION
**Pull request by lynnagara**

### CODE REVIEW

File: src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx [hunk 1]
```diff
@@ -231,3 +232,37 @@ const LightGray = styled.span`
...
+ * @returns {Void}
+ */
+export function downloadAsCsv(result) {
+  const {meta, data} = result;
+  const headings = meta.map(({name}) => name);
+
+  const csvContent = Papa.unparse({
+    fields: headings,
+    data: data.map(row => {
+      return headings.map(col => quoteUnsafeStrings(row[col]));
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx]: Is `papaparse` able to handle this?

Comment (lynnagara) [reply src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx]: I don't think so, I was hoping to not have to do this manually too

Comment (billyvg) [reply src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx]: What if we just quoted every field (I'm assuming papaparse would then handle escaping `"`)?

```
Set quotes to true to always enclose each field in quotes, or an array of true/false values correlating to specific to columns to force-quote. The character used to quote can be customized using quoteChar. The character used to escape the quoteChar within a field can be customized using escapeChar. The delimiter can be any valid delimiting character. The newline character(s) may also be customized. Setting header to false will omit the header row.
```

Comment (lynnagara) [reply src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx]: I wouldn't want to get a CSV like that

File: src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx [hunk 2]
```diff
@@ -231,3 +232,37 @@ const LightGray = styled.span`
...
+  const encodedDataUrl = encodeURI(`data:text/csv;charset=utf8,${csvContent}`);
+
+  window.location.assign(encodedDataUrl);
+}
+
+function quoteUnsafeStrings(value) {
+  const unsafeCharacterRegex = /^[\=\+\-\@]/;
+
+  if (typeof value === 'string' && `${value}`.match(unsafeCharacterRegex)) {
+    return `"${value}"`;
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx]: Let's insert a tab before the value: http://georgemauer.net/2017/10/07/csv-injection.html

**Review by billyvg — APPROVED** [review]

File: src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx [hunk 3]
```diff
@@ -231,3 +232,37 @@ const LightGray = styled.span`
...
+  const encodedDataUrl = encodeURI(`data:text/csv;charset=utf8,${csvContent}`);
+
+  window.location.assign(encodedDataUrl);
+}
+
+function quoteUnsafeStrings(value) {
+  const unsafeCharacterRegex = /^[\=\+\-\@]/;
+
+  if (typeof value === 'string' && `${value}`.match(unsafeCharacterRegex)) {
+    return `\t"${value}"`;
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx]: Can your data have `"` inside the fields? Right now `"` is not in your unsafe list and this quoting is not replacing the `"` with `""`.

Comment (lynnagara) [reply src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx]: Ok yeah, in that case maybe I'd also add this to any field that starts with `"=`, `"+`, `=-`, `=@`

Comment (lynnagara) [reply src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx]: Hmm, papa parse automatically escapes quotes though so maybe this is not required

Comment (markstory) [reply src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx]: You can prevent macros by prepending a `'` to fields that start with a macro byte.

**Comment by lynnagara** [comment]
Discussed with @markstory offline, we'll just add `'` in front of potential macros, papaparse already handles the rest of the escaping so we're good with everything else

**Pull request merged**